coderd/workspaceapps.go

```diff
@@ -73,45 +76,28 @@ func (api *API) workspaceApplicationAuth(rw http.ResponseWriter, r *http.Request
 		})
 		return
 	}
-	// Force the redirect URI to use the same scheme as the access URL for
-	// security purposes.
-	u.Scheme = api.AccessURL.Scheme
 
-	ok := false
-	if api.AppHostnameRegex != nil {
-		_, ok = httpapi.ExecuteHostnamePattern(api.AppHostnameRegex, u.Host)
+	u.Scheme, err = api.ValidWorkspaceAppHostname(ctx, u.Host, ValidWorkspaceAppHostnameOpts{
+		// Allow all hosts except primary access URL since we don't need app
+		// tokens on the primary dashboard URL.
+		AllowPrimaryAccessURL: false,
+		AllowPrimaryWildcard:  true,
+		AllowProxyAccessURL:   true,
+		AllowProxyWildcard:    true,
+	})
+	if err != nil {
+		httpapi.Write(ctx, rw, http.StatusInternalServerError, codersdk.Response{
+			Message: "Failed to verify redirect_uri query parameter.",
+			Detail:  err.Error(),
+		})
+		return
 	}
-
-	// Ensure that the redirect URI is a subdomain of api.Hostname and is a
-	// valid app subdomain.
-	if !ok {
-		proxy, err := api.Database.GetWorkspaceProxyByHostname(ctx, u.Hostname())
-		if xerrors.Is(err, sql.ErrNoRows) {
-			httpapi.Write(ctx, rw, http.StatusBadRequest, codersdk.Response{
-				Message: "The redirect_uri query parameter must be the primary wildcard app hostname, a workspace proxy access URL or a workspace proxy wildcard app hostname.",
-			})
-			return
-		}
-		if err != nil {
-			httpapi.Write(ctx, rw, http.StatusInternalServerError, codersdk.Response{
-				Message: "Failed to get workspace proxy by redirect_uri.",
-				Detail:  err.Error(),
-			})
-			return
-		}
-
-		proxyURL, err := url.Parse(proxy.Url)
-		if err != nil {
-			httpapi.Write(ctx, rw, http.StatusInternalServerError, codersdk.Response{
-				Message: "Failed to parse workspace proxy URL.",
-				Detail:  xerrors.Errorf("parse proxy URL %q: %w", proxy.Url, err).Error(),
-			})
-			return
-		}
-
-		// Force the redirect URI to use the same scheme as the proxy access URL
-		// for security purposes.
-		u.Scheme = proxyURL.Scheme
+	if u.Scheme == "" {
+		httpapi.Write(ctx, rw, http.StatusBadRequest, codersdk.Response{
+			Message: "Invalid redirect_uri.",
+			Detail:  "The redirect_uri query parameter must be the primary wildcard app hostname, a workspace proxy access URL or a workspace proxy wildcard app hostname.",
+		})
+		return
```

Comment:
Would the scheme not be `""` if it is a wildcard hostname?

Reply:
No, it uses the access URL's scheme

Reply:
Added a comment